src/boot/efi/devicetree.c

```diff
@@ -106,6 +106,125 @@ EFI_STATUS devicetree_install(struct devicetree_state *state, EFI_FILE *root_dir
                         MAKE_GUID_PTR(EFI_DTB_TABLE), PHYSICAL_ADDRESS_TO_POINTER(state->addr));
 }
 
+static const char* devicetree_get_compatible(const void *dtb) {
+        if ((uintptr_t) dtb % alignof(FdtHeader) != 0)
+                return NULL;
+
+        const FdtHeader *dt_header = ASSERT_PTR(dtb);
+
+        if (be32toh(dt_header->magic) != UINT32_C(0xd00dfeed))
+                return NULL;
+
+        uint32_t dt_size = be32toh(dt_header->total_size);
+        uint32_t struct_off = be32toh(dt_header->off_dt_struct);
+        uint32_t struct_size = be32toh(dt_header->size_dt_struct);
+        uint32_t strings_off = be32toh(dt_header->off_dt_strings);
+        uint32_t strings_size = be32toh(dt_header->size_dt_strings);
+        uint32_t end;
+
+        if (PTR_TO_SIZE(dtb) > SIZE_MAX - dt_size)
+                return NULL;
+
+        if (!ADD_SAFE(&end, strings_off, strings_size) || end > dt_size)
+                return NULL;
+        const char *strings_block = (const char *) ((const uint8_t *) dt_header + strings_off);
+
+        if (struct_off % sizeof(uint32_t) != 0)
+                return NULL;
+        if (struct_size % sizeof(uint32_t) != 0 ||
+            !ADD_SAFE(&end, struct_off, struct_size) ||
+            end > strings_off)
+                return NULL;
+        const uint32_t *cursor = (const uint32_t *) ((const uint8_t *) dt_header + struct_off);
+
+        size_t size_words = struct_size / sizeof(uint32_t);
+        size_t len, name_off, len_words, s;
+
+        for (size_t i = 0; i < end; i++) {
+                switch (be32toh(cursor[i])) {
+                case FDT_BEGIN_NODE:
+                        if (i >= size_words || cursor[++i] != 0)
+                                return NULL;
+                        break;
+                case FDT_NOP:
+                        break;
+                case FDT_PROP:
+                        /* At least 3 words should present: len, name_off, c (nul-terminated string always has non-zero length) */
+                        if (i + 3 >= size_words || cursor[++i] != 0)
+                                return NULL;
+                        len = be32toh(cursor[++i]);
+                        name_off = be32toh(cursor[++i]);
+                        len_words = DIV_ROUND_UP(len, sizeof(uint32_t));
+
+                        if (ADD_SAFE(&s, name_off, STRLEN("compatible")) &&
+                            s < strings_size && memcmp(strings_block + name_off, "compatible", STRLEN("compatible"))) {
```

Comment:
memory_startswith(strings_block + name_off, "compatible")

Reply:
there's strings_size, no?

Reply:
I guess it'll work. But don't we really want to only match `compatible\0`?

Reply:
Nope, it actually won't:
```
src/boot/efi/devicetree.c:160:107: error: passing argument 3 of 'memory_startswith' from incompatible pointer type [-Wincompatible-pointer-types]
  160 |                             s < strings_size && memory_startswith(strings_block + name_off, strings_size, "compatible")) {
      |                                                                                                           ^~~~~~~~~~~~
      |                                                                                                           |
      |                                                                                                           const char *
In file included from src/boot/efi/util.h:9,
                 from src/boot/efi/devicetree.c:5:
src/fundamental/string-util-fundamental.h:88:80: note: expected 'const sd_char *' {aka 'const short unsigned int *'} but argument is of type 'const char *'
   88 | static inline void *memory_startswith(const void *p, size_t sz, const sd_char *token) {
```

So it appears to be `char16_t` and doesn't look like what we want here.

Reply:
Should I keep it as is (`memcmp`) or go back to `streq8`?

Reply:
up to you.

other option: add memory_startswith8()

Reply:
Since this is nul-terminated string as per the spec, it makes sense to use `streq8` here.